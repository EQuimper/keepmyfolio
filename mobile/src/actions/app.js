// @flow

import type { Action } from '../types';

export function toggleTheme(): Action {
  return {
    type: 'app/TOGGLE_THEME',
  };
}

export function toggleSearchBar(): Action {
  return {
    type: 'app/TOGGLE_SEARCH_BAR',
  };
}
