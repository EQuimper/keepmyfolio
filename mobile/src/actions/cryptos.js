// @flow

import type { HoldingData } from '../types';

export function addNewHolding(coin: HoldingData) {
  return {
    type: 'cryptos/ADD_NEW_HOLDING',
    coin
  }
}
