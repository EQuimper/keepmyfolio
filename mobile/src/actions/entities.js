// @flow

import { CoinMarket } from '../utils/api';

import type { CoinMarketCapData } from '../types';

function getAllCoinMarketSuccess(data: CoinMarketCapData) {
  return {
    type: 'GET_ALL_COIN_MARKET_SUCCESS',
    data
  }
}

function getAllCoinMarketError(error: Error) {
  return {
    type: 'GET_ALL_COIN_MARKET_ERROR',
    error
  }
}

export function getAllCoinMarket() {
  return async (dispatch) => {
    dispatch({ type: 'GET_ALL_COIN_MARKET' });
    try {
      const { data } = await CoinMarket.getAll(10);
      return dispatch(getAllCoinMarketSuccess(data));
    } catch (error) {
      return dispatch(getAllCoinMarketError(error));
    }
  }
}
