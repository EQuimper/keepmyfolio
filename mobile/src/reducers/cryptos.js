// @flow

import { Record, Map } from 'immutable';

import type { Action, CryptosState, HoldingData } from '../types';

// TODO: Make sure the transaction it's a map

const StateRecord = Record({
  entities: new Map(),
  transactionId: 0,
});

function addNewHolding(state, action) {
  const _transactionId: number = state.get('transactionId') + 1;

  // const mapObj = Map(state.get('transactionId'), action.coin)

  const _newTransaction: Map<string, HoldingData> = Map({
    [state.get('transactionId')]: action.coin,
  });

  const _entities = state
    .get('entities')
    .mergeDeepIn([action.coin.cryptoId], _newTransaction);

  return state.set('transactionId', _transactionId).set('entities', _entities);
}

export default function cryptos(
  state: CryptosState = new StateRecord(),
  action: Action,
): CryptosState {
  switch (action.type) {
    case 'cryptos/ADD_NEW_HOLDING':
      return addNewHolding(state, action);
    default:
      return state;
  }
}
