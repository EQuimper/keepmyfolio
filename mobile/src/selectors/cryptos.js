// @flow

import createCachedSelector from 're-reselect';
import { Map } from 'immutable';

import type { State } from '../types';

import { moneyThousand } from '../utils/helpers/numbers';

const getAssetsEntities = (state: State, props) =>
  state.get('cryptos').entities.get(props.coin.cryptoId);

export const getAsset = createCachedSelector(
  [getAssetsEntities],
  asset => asset,
)((state, props) => props.coin.id);

const getPrice = (state: State, props) => parseFloat(props.coin.priceUsd);
const getPercentChange1h = (state: State, props) =>
  parseFloat(props.coin.percentChange1h);

export const getHolding = createCachedSelector(
  [getAssetsEntities],
  (entities: Map<string, Map<string, any>>) => {
    if (entities == null) {
      return 0;
    }

    const totalAmount: number = entities.reduce(
      (prev, current) => prev + parseFloat(current.get('amountOfCoin')),
      0,
    );

    return totalAmount;
  },
)((state, props) => props.coin.id);

export const getTotal = createCachedSelector(
  [getHolding, getPrice],
  (holding: ?number, price: number) => {
    if (!holding) {
      return null;
    }

    return moneyThousand(String(holding * price));
  },
)((state, props) => props.coin.id);

export const getAmountChange = createCachedSelector(
  [getHolding, getPrice, getPercentChange1h],
  (holding: ?number, price: number, percentChange: number) => {
    if (!holding) {
      return null;
    }

    if (Math.abs(percentChange) === 0) {
      return null;
    }

    const totalDollarUserHave: number = holding * price;

    return moneyThousand(String(totalDollarUserHave * percentChange / 100));
  },
)((state, props) => props.coin.id);
