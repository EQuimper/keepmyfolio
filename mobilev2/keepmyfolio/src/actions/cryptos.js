// @flow

import type { HoldingData, Action } from '../types';

export function addNewHolding(coin: HoldingData): Action {
  return {
    type: 'cryptos/ADD_NEW_HOLDING',
    coin
  }
}
