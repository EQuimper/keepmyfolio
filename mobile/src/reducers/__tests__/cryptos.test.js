// @noflow

import { addNewHolding } from '../../actions/cryptos';
import reducer, { StateRecord } from '../cryptos';

describe('Cryptos Reducer', () => {
  describe('#addNewHolding()', () => {
    const holding1 = {
      amountOfCoin: '2',
      cryptoId: 'bitcoin',
      dateBuy: new Date('Sat Sep 30 2017 12:49:28 GMT-0400 (EDT)'),
      id: 'Q3J5cHRvOmJpdGNvaW4=',
      name: 'Bitcoin',
      priceByCoin: '4325.7',
      priceTotalPay: '8651.40',
    };

    const holding2 = {
      amountOfCoin: '1',
      cryptoId: 'bitcoin',
      dateBuy: new Date('Sat Sep 30 2017 12:49:28 GMT-0400 (EDT)'),
      id: 'Q3J5cHRvOmJpdGNvaW4=',
      name: 'Bitcoin',
      priceByCoin: '4325.7',
      priceTotalPay: '4325.7',
    };

    const holding3 = {
      amountOfCoin: '20',
      cryptoId: 'neo',
      dateBuy: new Date('Sat Sep 30 2017 12:49:28 GMT-0400 (EDT)'),
      id: 'Q3J5cHRvOm5lbw==',
      name: 'NEO',
      priceByCoin: '32.2396',
      priceTotalPay: '644.79',
    };

    it('should add a holding', () => {
      const state = new StateRecord();
      expect(
        reducer(state, addNewHolding(holding1))
      ).toMatchSnapshot();
    });

    it('should add an existing type of coin a new holding', () => {
      let state = new StateRecord();
      state = reducer(state, addNewHolding(holding1));
      expect(state).toMatchSnapshot();
      state = reducer(state, addNewHolding(holding2));
      expect(state).toMatchSnapshot();
    })

    it('should add new coin on top of the old state', () => {
      let state = new StateRecord();
      state = reducer(state, addNewHolding(holding1));
      expect(state).toMatchSnapshot();
      state = reducer(state, addNewHolding(holding2));
      expect(state).toMatchSnapshot();
      state = reducer(state, addNewHolding(holding3));
      expect(state).toMatchSnapshot();
    })
  });
});
