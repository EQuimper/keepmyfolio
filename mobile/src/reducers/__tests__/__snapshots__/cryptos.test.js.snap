// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cryptos Reducer #addNewHolding() should add a holding 1`] = `
Immutable.Record {
  "entities": Immutable.Map {
    "bitcoin": Immutable.Map {
      "0": Immutable.Map {
        "amountOfCoin": "2",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "8651.40",
      },
    },
  },
  "transactionId": 1,
}
`;

exports[`Cryptos Reducer #addNewHolding() should add an existing type of coin a new holding 1`] = `
Immutable.Record {
  "entities": Immutable.Map {
    "bitcoin": Immutable.Map {
      "0": Immutable.Map {
        "amountOfCoin": "2",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "8651.40",
      },
    },
  },
  "transactionId": 1,
}
`;

exports[`Cryptos Reducer #addNewHolding() should add an existing type of coin a new holding 2`] = `
Immutable.Record {
  "entities": Immutable.Map {
    "bitcoin": Immutable.Map {
      "0": Immutable.Map {
        "amountOfCoin": "2",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "8651.40",
      },
      "1": Immutable.Map {
        "amountOfCoin": "1",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "4325.7",
      },
    },
  },
  "transactionId": 2,
}
`;

exports[`Cryptos Reducer #addNewHolding() should add new coin on top of the old state 1`] = `
Immutable.Record {
  "entities": Immutable.Map {
    "bitcoin": Immutable.Map {
      "0": Immutable.Map {
        "amountOfCoin": "2",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "8651.40",
      },
    },
  },
  "transactionId": 1,
}
`;

exports[`Cryptos Reducer #addNewHolding() should add new coin on top of the old state 2`] = `
Immutable.Record {
  "entities": Immutable.Map {
    "bitcoin": Immutable.Map {
      "0": Immutable.Map {
        "amountOfCoin": "2",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "8651.40",
      },
      "1": Immutable.Map {
        "amountOfCoin": "1",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "4325.7",
      },
    },
  },
  "transactionId": 2,
}
`;

exports[`Cryptos Reducer #addNewHolding() should add new coin on top of the old state 3`] = `
Immutable.Record {
  "entities": Immutable.Map {
    "bitcoin": Immutable.Map {
      "0": Immutable.Map {
        "amountOfCoin": "2",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "8651.40",
      },
      "1": Immutable.Map {
        "amountOfCoin": "1",
        "cryptoId": "bitcoin",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOmJpdGNvaW4=",
        "name": "Bitcoin",
        "priceByCoin": "4325.7",
        "priceTotalPay": "4325.7",
      },
    },
    "neo": Immutable.Map {
      "2": Immutable.Map {
        "amountOfCoin": "20",
        "cryptoId": "neo",
        "dateBuy": 2017-09-30T16:49:28.000Z,
        "id": "Q3J5cHRvOm5lbw==",
        "name": "NEO",
        "priceByCoin": "32.2396",
        "priceTotalPay": "644.79",
      },
    },
  },
  "transactionId": 3,
}
`;

exports[`Cryptos Reducer Initial State should return the initialState on default 1`] = `
Immutable.Record {
  "entities": Immutable.Map {},
  "transactionId": 0,
}
`;
